### PR TITLE
Check whether or not the given time is valid

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes:  
https://github.com/ballerina-platform/ballerina-standard-library/issues/2320
https://github.com/ballerina-platform/ballerina-standard-library/issues/2497
## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: ballerina/tests/schedulerTest.bal [hunk 1]
```diff
@@ -600,3 +599,17 @@ isolated function testScheduleJobsWithNegativeInterval() returns error? {
...
 }
+
+@test:Config {
+    groups: ["FrequencyJob", "negative"]
+}
+isolated function testScheduleJobsWithInvalidStartTime() returns error? {
+    time:Civil startTime = {"utcOffset":{"hours":0,"minutes":0},"timeAbbrev":"Z","dayOfWeek":1,"year":2021,"month":12,
+                            "day":13,"hour":7,"minute":25,"second":40.893987};
+    JobId|Error output = scheduleJobRecurByFrequency(new Job23(), 4, startTime = startTime);
+    if(output is Error) {
```
Comment (MadhukaHarith92) [inline ballerina/tests/schedulerTest.bal:610]:
```suggestion
    if output is Error {
```

File: ballerina/tests/schedulerTest.bal [hunk 2]
```diff
@@ -600,3 +599,17 @@ isolated function testScheduleJobsWithNegativeInterval() returns error? {
...
+        test:assertTrue(output.message().includes("Invalid time"), output.message());
+    } else {
+        test:assertFail("Test failed.");
```
Comment (MadhukaHarith92) [inline ballerina/tests/schedulerTest.bal:613]:
```suggestion
        test:assertFail("scheduleJobRecurByFrequency did not fail with invalid start time");
```

**Review by daneshk — APPROVED** [review]

File: ballerina/commons.bal [hunk 3]
```diff
@@ -64,7 +64,11 @@ public isolated function getTimeInMillies(time:Civil time) returns int|Error {
     }
     time:Utc|time:Error utc = time:utcFromCivil(time);
     if utc is time:Utc {
-        return <int> decimal:round((<decimal>utc[0] + utc[1]) * 1000);
+        if utc > time:utcNow() {
+            return <int> decimal:round((<decimal>utc[0] + utc[1]) * 1000);
+        }
+        string timeInString = time.toString();
+        return error Error(string `Invalid time: ${timeInString}.`);
```
Comment (daneshk) [inline ballerina/commons.bal:71]: Can't we just pass the `time` value to the string template?

Comment (kalaiyarasiganeshalingam) [reply ballerina/commons.bal:71]: The string template expects int|float|decimal|string|boolean. But here, time is a record. So we have to convert to the string.

File: ballerina/tests/scheduler_test.bal [hunk 4]
```diff
@@ -597,3 +599,17 @@ isolated function testScheduleJobsWithNegativeInterval() returns error? {
         test:assertFail("Test failed.");
     }
 }
+
+@test:Config {
+    groups: ["FrequencyJob", "negative"]
+}
+isolated function testScheduleJobsWithInvalidStartTime() returns error? {
+    time:Civil startTime = {"utcOffset":{"hours":0,"minutes":0},"timeAbbrev":"Z","dayOfWeek":1,"year":2021,"month":12,
+                            "day":13,"hour":7,"minute":25,"second":40.893987};
```
Comment (daneshk) [inline ballerina/tests/scheduler_test.bal:608]: Is this correctly formatted?

File: ballerina/commons.bal [hunk 5]
```diff
@@ -64,7 +64,10 @@ public isolated function getTimeInMillies(time:Civil time) returns int|Error {
     }
     time:Utc|time:Error utc = time:utcFromCivil(time);
     if utc is time:Utc {
-        return <int> decimal:round((<decimal>utc[0] + utc[1]) * 1000);
+        if utc > time:utcNow() {
+            return <int> decimal:round((<decimal>utc[0] + utc[1]) * 1000);
+        }
+        return error Error(string `Invalid time: ${time.toString()}.`);
```
Comment (daneshk) [inline ballerina/commons.bal:70]:
```suggestion
        return error Error(string `Invalid time: ${time.toString()}. Scheduled time should be greater than the current time`);
```

File: ballerina/commons.bal [hunk 6]
```diff
@@ -64,7 +64,11 @@ public isolated function getTimeInMillies(time:Civil time) returns int|Error {
     }
     time:Utc|time:Error utc = time:utcFromCivil(time);
     if utc is time:Utc {
-        return <int> decimal:round((<decimal>utc[0] + utc[1]) * 1000);
+        if utc > time:utcNow() {
+            return <int> decimal:round((<decimal>utc[0] + utc[1]) * 1000);
+        }
+        return error Error(
+                    string `Invalid time: ${time.toString()}.Scheduled time should be greater than the current time`);
```
Comment (daneshk) [inline ballerina/commons.bal:71]:
```suggestion
                    string `Invalid time: ${time.toString()}. Scheduled time should be greater than the current time`);
```

**Review by daneshk — APPROVED** [review]

**Pull request merged**